chalk-solve/src/clauses.rs

```diff
@@ -952,3 +953,109 @@ impl ToProgramClauses for AssociatedTyDatum {
         );
     }
 }
+
+/// Returns a set of program clauses that could possibly match
+/// `goal`. This can be any superset of the correct set, but the
+/// more precise you can make it, the more efficient solving will
+/// be.
+pub fn program_clauses_that_could_match(
+    program: &dyn RustIrDatabase,
+    goal: &DomainGoal,
+    vec: &mut Vec<ProgramClause>,
+) {
+    let mut clauses = vec![];
+    match goal {
+        DomainGoal::Holds(WhereClause::Implemented(trait_ref)) => {
+            program
+                .trait_datum(trait_ref.trait_id)
+                .to_program_clauses(program, &mut clauses);
```

Comment:
Something seems fishy here -- I'm surprised this never has to lower the *impls* of the given trait

Reply:
Indeed, in the [rust source code](https://github.com/rust-lang/rust/blob/3d720d7e617442aaf65df0910e8c9f0fdb0cf31c/src/librustc_traits/chalk_context/program_clauses/mod.rs#L72-L76), we do seem to lower the rules from the impl, although in a mildly confusing way.

Reply:
So I think we need something like

```rust
for impl_id in program.impls_for_trait(trait_ref.trait_id) {
    program.impl_datum(impl_id).to_program_clauses(program, &mut clauses);
}
```

Reply:
Actually i'm a bit confused by the rustc source, maybe @scalexm can explain. The comment says:

```rust
                // These come from:
                // * implementations of the trait itself (rule `Implemented-From-Impl`)
                // * the trait decl (rule `Implemented-From-Env`)
```

but the source seems to (a) lower the associated items from an impl and (b) not the trait declaration itself. I haven't looked more deeply though and I'm sure it all works out.